src/main/java/bio/terra/cli/context/utils/Logger.java

```diff
@@ -0,0 +1,184 @@
+      throw new IllegalArgumentException("Maximum number of log files must be >= 1");
+    }
+    rollingPolicy.setMinIndex(1);
+    rollingPolicy.setMaxIndex(MAX_NUM_FILES - 1);
+    rollingFileAppender.setRollingPolicy(rollingPolicy);
+
+    // make sure to start the policies after the cross-references between the policy and appender
+    // have been set
+    triggeringPolicy.start();
+    rollingPolicy.start();
+
+    setupEncoderAndFilter(
+        rollingFileAppender, loggerContext, globalContext.fileLoggingLevel.getLogLevelImpl());
+    rollingFileAppender.start();
+
+    // build the console appender
+    ConsoleAppender consoleAppender = new ConsoleAppender();
+    consoleAppender.setName("ConsoleAppender");
+    consoleAppender.setContext(loggerContext);
+    setupEncoderAndFilter(
+        consoleAppender, loggerContext, globalContext.consoleLoggingLevel.getLogLevelImpl());
+    consoleAppender.start();
+
+    // on the root logger, clear any existing appenders and attach the two created above
+    ch.qos.logback.classic.Logger rootLogger = loggerContext.getLogger(ROOT_LOGGER_NAME);
+    rootLogger.detachAndStopAllAppenders();
+    rootLogger.addAppender(rollingFileAppender);
+    rootLogger.addAppender(consoleAppender);
+
+    // if a process is too short-lived, then the policy may not check if it should rollover.
+    // each CLI command invocation is a new process (i.e. new JVM) and may be very short-lived.
+    // so, include an additional manual check here on startup, to see if we should roll over the
+    // file based on its size.
+    File activeFile = new File(rollingPolicy.getActiveFileName());
+    if (activeFile.length() > MAX_FILE_SIZE) {
+      rollingFileAppender.rollover();
+    }
+
+    // StatusPrinter.print(loggerContext); // helpful for debugging
+  }
+
+  /**
+   * Helper method to setup the encoder and filter for the given appender. This setup is common to
+   * all appenders.
+   *
+   * @param appender appender object to setup
+   * @param loggerContext global logger context object
+   * @param loggingLevel logging level particular to this appender
+   */
+  private static void setupEncoderAndFilter(
+      OutputStreamAppender appender, LoggerContext loggerContext, Level loggingLevel) {
+    PatternLayoutEncoder encoder = new PatternLayoutEncoder();
+    encoder.setContext(loggerContext);
+    encoder.setPattern(LOG_FORMAT);
+    encoder.start();
+    appender.setEncoder(encoder);
+
+    // filter out any logs that are below the fileLoggingLevel specified in the global context
+    appender.clearAllFilters();
+    appender.addFilter(
+        new Filter<ILoggingEvent>() {
+          @Override
+          public FilterReply decide(final ILoggingEvent event) {
+            return event.getLevel().isGreaterOrEqual(loggingLevel)
+                ? FilterReply.ACCEPT
+                : FilterReply.DENY;
+          }
+        });
+  }
+
+  /**
+   * Wrapper class around the ch.qos.logback.classic.Level class.
```

Comment:
[No action / side comment] I *think* I understand the intent here. But the only use case I can envision for this is in GlobalContext to configure the logback log level (either as part of a stored configuration or via command-line arguments). In which case, it's not too inappropriate to say that GlobalContext should know that our implementation is logback, because it's directly configuring the logback system.

Is there some other scenario you're protecting against, where these levels become more widely-used throughout the codebase?

Or, maybe the concern is more about how these values might get serialized alongside the GlobalContext? In which case a small note about serialization control might help clarify the need.

Reply:
Without this wrapper class, we'd leak logger implementation into:
- GlobalContext, including the serialized version persisted in a file.  ch.qos.logback.classic.Level is an object, not a string, and I didn't want this file to include logger implementation-specific object structure. That would make it harder to change later (i.e. because if we replace the cho.qos library with another one, then we won't have the appropriate  Level class any more to do the deserialization)
- (Upcoming) Config command to set the log level(s), including displaying the possible values.

Added a note about this to the code.